### ISSUE TITLE
Support send notification to Slack, RocketChat etc. using Webhooks

### ISSUE THREAD

**Issue by jmcshane** [brotandgames/ciao#1]
It appears that the notification integration is set here:

https://github.com/brotandgames/ciao/blob/master/app/models/check.rb#L52

It would be great to be able to send a notification to a slack channel instead of an email based on specific configuration. I would be interested in helping here using https://github.com/slack-ruby/slack-ruby-client.

Is this something the project is willing to support?

**Comment by mterron** [comment]
Or a webhook? That is a more flexible solution for integration.

**Comment by brotandgames** [comment]
@jmcshane @mterron Thanks for the comments.

Support for webhooks is on the roadmap.

@jmcshane Mind if we integrate Slack using webhooks? If you are eager to help us with this you're welcome.

**Comment by jmcshane** [comment]
@brotandgames that's great.

Definitely willing to jump in a bit here. I like the simplicity of the way that your email integration is set up and that can be more challenging given the structure of a webhook. What would you think about a default JSON format with something that can be overridden based on a configuration file in the image? Then, an environment var could trigger webook (and disable email) like ENABLE_WEBHOOK_NOTIFICATION, reading a set of environment variables to trigger the HTTP request.

**Comment by brotandgames** [comment]
Shoud it be allowed to use more than one Notification target like Mail, Webhook::Slack or Webhook::RocketChat at the same time? Our opinion: yes.

As of now if you don't specify a `SMTP_ADDRESS` the Mailer configuration is not read thus no e-mails are sent. This approach would be preferred in the context of Webhooks.

Webhooks should somehow be configurable through ENV variables to stay in the [12-factor app context](https://12factor.net/config).

Furthermore there should be no updates on the app if we want to support another notification target since it's a webhook: calling an endpoint with a payload.

What about having the following approach:

2 ENV variables per Webhook like:
````
CIAO_WEBHOOK_ENDPOINT_1="https://webhook.slack.com/***/***"
CIAO_WEBHOOK_PAYLOAD_1="{foo=bar}"

CIAO_WEBHOOK_ENDPOINT_2="https://webhook.rocketchat.com/***/***"
CIAO_WEBHOOK_PAYLOAD_2="{foo=bar}"

etc.
````
The app reads the ENV vars on startup and configures the Notifier which is then triggered by a status change.

**Comment by brotandgames** [comment]
Tested this a little bit:

`CIAO_WEBHOOK_PAYLOAD_*` ENV variable has to be a valid JSON one-liner wrapped in single quotes like `'{"text":"Example message"}'`.
````ruby
require 'net/http'
require 'json'
require 'uri'

# Following ENV vars have to be set prior to executing this script
# export CIAO_WEBHOOK_ENDPOINT_1=https://chat.yourhost.com/hooks/*****
# export CIAO_WEBHOOK_PAYLOAD_1='{"username":"Brot & Games","icon_url":"https://avatars0.githubusercontent.com/u/43862266?s=400&v=4","text":"Example message","attachments":[{"title":"Rocket.Chat","title_link":"https://rocket.chat","text":"Rocket.Chat, the best open source chat","image_url":"/images/integration-attachment-example.png","color":"#764FA5"}]}'

endpoint = ENV["CIAO_WEBHOOK_ENDPOINT_1"]
payload = ENV["CIAO_WEBHOOK_PAYLOAD_1"]

uri = URI.parse(endpoint)
http = Net::HTTP.new(uri.host, uri.port)
http.use_ssl = true if uri.scheme == "https"

request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
request.body = payload

response = http.request(request)
p response

````

This has been tested with RocketChat (with the above payload `$CIAO_WEBHOOK_PAYLOAD_1`)

````
$ ruby webhoook_test.rb
#<Net::HTTPOK 200 OK readbody=true>
````

@jmcshane Do you currently have a Slack channel where you could test this?

**Comment by tareksamni** [comment]
I can implement this in a simple dynamic way where the user will be able to CRUD multiple webhooks and the payload structure (basic templating) from the web UI. I can deliver it today evening! 

Should I go forward or any opinions?

**Comment by brotandgames** [comment]
@tareksamni The basic idea of ciao is to keep everything configurable per ENV variables.
The basic approach of having ENV variables like:
````
WEBHOOK_ENDPOINT_ROCKETCHAT
WEBHOOK_PAYLOAD_ROCKETCHAT

or

WEBHOOK_ENDPOINT_SLACK
WEBHOOK_PAYLOAD_SLACK

````
would be desirable.

Keeping ciao's configuration *stateless* is keeping it simple.

**Comment by tareksamni** [comment]
Good, then we go with ENV variables `WEBHOOK_ENDPOINT_1`, `WEBHOOK_PAYLOAD_TEMPLATE_1`, `WEBHOOK_ACTION_1`.

Something like:

```
WEBHOOK_ENDPOINT_1: https://myendpoint.com/webhook
WEBHOOK_ACTION_1: POST
WEBHOOK_PAYLOAD_TEMPLATE_1: '{ "status": "#{{status}}" }' # I would use http://mustache.github.io/#demo for simple payload templating!
```

You can define multiple webhooks with this regex: `^WEBHOOK_(ACTION||ENDPOINT||PAYLOAD_TEMPLATE)_[0-9]+`

What do you think?

**Comment by brotandgames** [comment]
Sound good.

We need an initializer `notification_webhooks.rb` that 

- reads all `WEBHOOK_ENDPOINT_` variables, 
- finds the corresponding  `WEBHOOK_PAYLOAD_` and 
- instantiates an `WebhookNotification < Notification` object

Something like this:

`lib/ciao/notification.rb`
````
module Ciao

  class Notification

    def initialize
    end

    def send
    end

  end

  class WebhookNotification < Notification
    # HTTP endpoint
    attr_reader :endpoint

    # Payload (JSON string)
    attr_reader :payload

    def initialize(endpoint, payload)
      @endpoint = endpoint
      @payload = payload
    end

    def send
      uri = URI.parse(@endpoint)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true if uri.scheme == "https"

      request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
      request.body = @payload

      response = http.request(request)
      p response
    end

    def to_s
      puts "Endpoint: #{@endpoint}"
      puts "Payload: #{@payload}"
    end

  end

end
````

`config/initializers/notifications.rb`
````
# Notifications
notifications = []

# Webhooks
# Read all ENV variables starting with CIAO_WEBHOOK_ENDPOINT_*
# and initialize appropriate class
# Test
# export CIAO_WEBHOOK_ENDPOINT_1=https://chat.yourhost.net/*****
# export CIAO_WEBHOOK_PAYLOAD_1='{"username":"Brot & Games","icon_url":"https://avatars0.githubusercontent.com/u/43862266?s=400&v=4","text":"Example message","attachments":[{"title":"Rocket.Chat","title_link":"https://rocket.chat","text":"Rocket.Chat, the best open source chat","image_url":"/images/integration-attachment-example.png","color":"#764FA5"}]}'


# endpoint = ENV["CIAO_WEBHOOK_ENDPOINT_1"]
# payload = ENV["CIAO_WEBHOOK_PAYLOAD_1"]
# webhook = Ciao::WebhookNotification.new(endpoint, payload)
# notifications << webhook
#
#
# NOTIFICATIONS = notifications
````

**Comment by tareksamni** [comment]
Yes, exactly what I had in mind :) but with slight templating abilities to the `PAYLOAD`. To generically support all webhooks based notifications so we do not have to implement different JSON structures for different systems.

**Comment by brotandgames** [comment]
Perfect.

**Comment by brotandgames** [comment]
@tareksamni @gnomus @mterron @jmcshane  Is this a good timing to open a chat like gitter.im for this project?

**Comment by tareksamni** [comment]
YES YES YES! would be way easier to discuss features, implementation strategies, etc.

**Comment by brotandgames** [comment]
Here we go: https://gitter.im/brotandgames/ciao

**Comment by brotandgames** [comment]
@tareksamni 
Why use mustache and not substitute a payload like this:
WEBHOOK_PAYLOAD_* contains `__check_name__`, `__status_after__`, `status_before__` and so on and is just substituted (gsub) before send with current values. 

All WebhookNotification are instantiated in the initializer and saved in an Array (notifications).
On status change go through array and each.send|notify?

**Comment by brotandgames** [comment]
@tareksamni Thank you for your great work!

Closed by #25.
